File: src/main/scala/packt/spark/ZipfValidation.scala
package packt.spark

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import packt.HelperScala

object ZipfValidation {

  def main(args: Array[String]): Unit = {
    val session: SparkSession = HelperScala.createSession(2, "Zipf Validation")
    val lines: RDD[String] = session.sparkContext.textFile(HelperScala.novellaLocation)

    val countsPerToken: RDD[(String, Int)] = lines
      .flatMap(_.toLowerCase.split("\\s+")) // split on whitespace
      .map(word => (word.replaceAll("(^[^a-z0-9]+|[^a-z0-9]+$)", ""), 1)) // removing punctuation
      .reduceByKey((count1, count2) => count1 + count2, 2)

    val sortedCountsToken = countsPerToken.map(_.swap) // swap is equivalent to `pair =>(pair._2, pair._1)`
      .sortByKey(false)

    sortedCountsToken.saveAsTextFile("./sortedcounts")
  }

}

File: src/main/python/packt/solutions/zipf_validation.py
from pyspark.rdd import RDD
from pyspark.sql import SparkSession
from packt.helper_python import create_session, novella_location
from operator import add
import re

if __name__ == "__main__":
    session: SparkSession = create_session(2, "Zipf validation")
    lines: RDD = session.sparkContext.textFile(novella_location)

    counts_per_token = lines \
        .flatMap(lambda line: line.lower().split()) \
        .map(lambda word: (re.sub(r"(^[^a-z0-9]+|[^a-z0-9]+$)", '', word), 1)) \
        .reduceByKey(add)

    sorted_counts = counts_per_token.map(lambda tuple: (tuple[1], tuple[0])) \
        .sortByKey(ascending=False)

    sorted_counts.saveAsTextFile("./sortedcounts")